core-api/src/main/java/org/silverpeas/core/persistence/jdbc/sql/JdbcSqlQuery.java

```diff
@@ -330,41 +375,95 @@ public JdbcSqlQuery in(Collection<?> parameters) {
     addListOfParameters(Arrays.asList(parameters), true);
     return this;
   }
 
   /**
-   * Centralization in order to build easily a SQL values clause.
+   * Centralization in order to build easily a SQL in clause.
+   * @param parameters the parameters to append to the given SQL query.
    * @return the instance of {@link JdbcSqlQuery} that represents the SQL query.
    */
-  private JdbcSqlQuery valuesForInsert() {
+  public JdbcSqlQuery notIn(Collection<?> parameters) {
+    sqlQuery.append(NOT_IN_OPERATOR);
+    addListOfParameters(parameters, true);
+    return this;
+  }
+
+  /**
+   * Centralization in order to build easily a SQL in clause.
+   * @param parameters the parameters to append to the given SQL query.
+   * @return the instance of {@link JdbcSqlQuery} that represents the SQL query.
+   */
+  public JdbcSqlQuery notIn(Object... parameters) {
+    sqlQuery.append(NOT_IN_OPERATOR);
+    addListOfParameters(Arrays.asList(parameters), true);
+    return this;
+  }
+
+  /**
+   * Centralization in order to build easily a SQL values clause.
+   */
+  private void valuesForInsert() {
     sqlQuery.append(") values");
     addListOfParameters(allParameters, false);
-    return this;
   }
 
   /**
    * Centralization in order to build easily a SQL values clause.
-   * @return the instance of {@link JdbcSqlQuery} that represents the SQL query.
    */
-  private JdbcSqlQuery finalizeTableCreation() {
+  private void finalizeTableCreation() {
     sqlQuery.append(")");
-    return this;
   }
 
-  private JdbcSqlQuery addListOfParameters(Collection<?> parameters,
+  private void addListOfParameters(Collection<?> parameters,
       final boolean addToParameters) {
-    String params = "";
     if (parameters != null) {
-      params = parameters.stream().map(p -> "?").collect(Collectors.joining(","));
+      // Oracle has a hard limitation with SQL lists with 'in' clause: it cannot take more than 1000
+      // elements. So we split it in several SQL lists so that they contain less than 1000 elements.
+      final int threshold = 1000;
+      int end = sqlQuery.lastIndexOf(NOT_IN_OPERATOR);
+      String negation = " NOT ";
+      if (end < 0) {
+        end = sqlQuery.lastIndexOf(IN_OPERATOR);
+        negation = "";
+      }
+      if (end > -1 && parameters.size() > threshold) {
+        int from = end - 1;
+        while (sqlQuery.charAt(from) != ' ') {
+          from--;
+        }
+        String currentSqlPart = sqlQuery.substring(from + 1, end);
+        sqlQuery.delete(from + 1, sqlQuery.length()).append(negation).append("(");
+
+        List<?> listOfParameters = new ArrayList<>(parameters);
+        for (int i = 0; i < listOfParameters.size(); i += threshold) {
+          String params = sublistOfParameters(i, i + threshold, listOfParameters);
+          sqlQuery.append(currentSqlPart).append(" IN (").append(params).append(") OR ");
+        }
+        sqlQuery.replace(sqlQuery.length() - 4, sqlQuery.length(), ")");
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)